sql/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -49,16 +49,22 @@
 
     private final RowShardResolver rowShardResolver;
     private final List<? extends CollectExpression<Row, ?>> expressions;
+    private final List<? extends CollectExpression<Row, ?>> sourceInfoExpressions;
     private final Function<String, TItem> itemFactory;
     private final Supplier<String> indexNameResolver;
     private final ClusterService clusterService;
     private final boolean autoCreateIndices;
+    private final BiConsumer<ShardedRequests, String> itemFailureConsumer;
+    private final Function<ShardedRequests, Boolean> possibleSourceUriProcessor;
 
     GroupRowsByShard(ClusterService clusterService,
                      RowShardResolver rowShardResolver,
                      Supplier<String> indexNameResolver,
                      List<? extends CollectExpression<Row, ?>> expressions,
+                     List<? extends CollectExpression<Row, ?>> sourceInfoExpressions,
                      Function<String, TItem> itemFactory,
+                     BiConsumer<ShardedRequests, String> itemFailureConsumer,
+                     Function<ShardedRequests, Boolean> possibleSourceUriProcessor,
```

Comment:
Would prefer if these were not Consumer/Function and we would directly call methods on the `UpsertResultContext`. This is an indirection which can be a bit tricky to understand.